💻 feat: Deeper MCP UI integration in the Chat UI

samuelpath opened the pull request:
> [!NOTE]  
> Most lines of code changed in this PR are tests. Advice for reviewer: first toggle out all test files to focus on the main changes in the logic, since there are not that many of them. And only then can you review the tests.

## Summary

The past weeks, we shipped the following 3 PRs to integrate [MCP UI](https://mcpui.dev/) with LibreChat:
- https://github.com/danny-avila/LibreChat/pull/9299
- https://github.com/danny-avila/LibreChat/pull/9418
- https://github.com/danny-avila/LibreChat/pull/9472 (replaced by this one https://github.com/danny-avila/LibreChat/pull/9581)

These PRs made it such that whenever a tool response would send back a UI Resource, it would be displayed in the tool response detail section, as such:

<img width="713" height="689" alt="image" src="https://github.com/user-attachments/assets/55c4f7f4-30f5-46e4-b75e-02b8c7f429a5" />

This was only a first step.

The goal of the current PR is to allow the LLM to integrate UI Resource elements directly as part of the chat response. Here is what we would have now:

<img width="834" height="593" alt="image" src="https://github.com/user-attachments/assets/e7507066-6e0a-4cc6-a150-b854a216038e" />

Or this:

<img width="670" height="409" alt="image" src="https://github.com/user-attachments/assets/57d7f047-68cd-4473-8f95-35e002e70593" />

Or when the chat is able to aggregate UI Resources from multiple tool responses:

<img width="1402" height="1348" alt="image" src="https://github.com/user-attachments/assets/0d808a03-0c26-4bfe-9e48-aa08738f208a" />

You can now interact with UI Resources. When you click on an action for which the client should be aware, a message is posted to the client and can be picked up by the LLM.

For instance here I clicked on "Add to cart" in one of the product cards:

<img width="737" height="659" alt="image" src="https://github.com/user-attachments/assets/ab340297-4d93-41de-a8b9-07e5e97b268d" />

We see that LibreChat gets the message from the UI Resource, interprets it. For now, it's as if the UI Resource is impersonating the user. I will explain below how we intend to improve it in a follow-up PR.

The LLM then calls another tool call that flows from the user action (click on add to cart on a given product, product added to the cart, cart displayed):

<img width="697" height="545" alt="image" src="https://github.com/user-attachments/assets/db9b2c25-c174-44e5-978c-c313e0300be3" />

### Key changes made

**1. Plugin-Based Architecture for MCP UI Resources**
  - **New Plugin System**: Created mcpUIResourcePlugin that processes markdown text and converts UI resource markers (like ui0, ui1,2,3) into proper React components
  - **Inline UI Rendering**: UI resources now render directly inline within markdown content instead of being displayed separately

**2. Data Structure Improvements**
  - **Array Format**: Changed UI resources from object format {0: resource, 1: resource} to array format [resource1, resource2] for better indexing and iteration
  - **Simplified Processing**: This makes it easier to handle multiple resources and carousel displays

**3. New Components Created**
  - **MCPUIResource**: Renders individual UI resources based on markdown markers and message context
  - **MCPUIResourceCarousel**: Handles multiple UI resources in carousel format when multiple indices are specified
  - **Improved UIResourceCarousel**: Enhanced with proper action handling using handleUIAction

**4. Enhanced Integration Points**
  - **Markdown Processing**: Integrated the plugin into the main Markdown.tsx component to process UI resource markers
  - **Tool Context**: Added helpful context in handleTools.js:346 that instructs AI models how to properly format UI resource markers for inline display
  - **Action Handling**: Unified UI action handling through handleUIAction utility that properly submits messages when users interact with UI components

**5. Better User Experience**
  - **Inline Display**: UI resources now appear exactly where the AI places them in the text (e.g., "Here's the product details ui0")
  - **Carousel Support**: Multiple resources can be displayed together using comma-separated indices (e.g., ui0,1,2)
  - **Interactive Elements**: Users can now interact with UI components and their actions properly trigger new messages

 **6. Comprehensive Testing**
  - Added extensive test coverage for the new components and plugin functionality
  - Tests cover edge cases, error handling, and proper integration with message context

The overall improvement transforms MCP UI from a separate attachment display to a seamless, inline experience where AI models can precisely control where and how UI resources appear within their responses.

### Samples of improvement follow-ups to come

- Have a special persona in the chat flow for messages sent by a UI resource user action to the LLM
- Add a user setting to show or hide the messages sent by a UI resource user action to the LLM
- Handle hints sent by the MCP server response in the metadata about how to best display the UI Resource

## Change Type

Please delete any irrelevant options.

- [x] New feature (non-breaking change which adds functionality)
- [x] This change requires a documentation update (we'll do the documentation at the end of the iterations)

## Testing

Here are the MCP servers I use for my testing:

```yaml
mcpServers:
  mcp-ui-allbirds-tool-call:
    type: 'streamable-http'
    url: https://mcpstorefront.com/?storedomain=allbirds.com

  mcp-ui-allbirds-intent:
    type: 'streamable-http'
    url: https://mcpstorefront.com/?store=allbirds.com&style=default

  mcp-ui-allbirds-prompt:
    type: 'streamable-http'
    url: https://mcpstorefront.com/?store_domain=allbirds.com

  mcp-aharvard:
    type: 'streamable-http'
    url: https://mcp-aharvard.netlify.app/mcp
```

For the storefront servers, you can enter a prompt like: "looking for men sneakers", and then try to play with the UI elements displayed.

For the `mcp-aharvard` server, you can ask what the weather is like in some city.

I've shown above what kind of result you should be seeing.

### **Test Configuration**:

## Checklist

Please delete any irrelevant options.

- [x] My code adheres to this project's style guidelines
- [x] I have performed a self-review of my own code
- [x] I have commented in any complex areas of my code
- [x] My changes do not introduce new warnings
- [x] I have written tests demonstrating that my changes are effective or that my feature works
- [x] Local unit tests pass with my changes

danny-avila commented on client/src/components/Chat/Messages/Content/ToolCall.tsx:
```diff
@@ -99,6 +100,13 @@ export default function ToolCall({
 
+  // Store UI resources when output is available
+  useEffect(() => {
+    if (output && attachments && !isSubmitting) {
```
instead of adding this `useEffect`, can you leverage the existing `useAttachments` hook? 

`client/src/hooks/Messages/useAttachments.ts`

danny-avila commented on packages/api/src/mcp/utils.ts:
```diff
@@ -45,3 +45,44 @@ export function sanitizeUrlForLogging(url: string | URL): string {
   }
 }
+
+export const uiResourcesInstructions = `The tool response contains UI resources (i.e. Resource URI starting with "ui://"),
```
I think it's unnecessary to couple this with artifacts. We can easily show the UI resources adjacent to the tool call, or even in place of the tool call, without needing the LLM to "display" it

danny-avila commented on client/src/components/MCPUIResource/plugin.ts:
```diff
@@ -0,0 +1,85 @@
+import type { Node } from 'unist';
+import type { UIResourceNode } from './types';
+
+export const UI_RESOURCE_MARKER = '\ue205';
```
I would not recommend using this as a marker, and instead use `\ui` or something similar, especially since the marker does not vary from `\ue205`.

As long as the plugin correctly falls back to `\ui` or the chosen marker, if it has no corresponding resource, to avoid edge cases where the AI writes the marker for purposes not related to UI resources.

Using "unicode-like" markers was a convention introduced by ChatGPT for their web search, which was a large inspiration for LC's UI/LLM interaction

plgodin committed:
💻 feat: deeper MCP UI integration in the chat UI using plugins

---------

Co-authored-by: Samuel Path <samuel.path@shopify.com>
Co-authored-by: Pierre-Luc Godin <pierreluc.godin@shopify.com>

plgodin committed:
💻 refactor: Migrate MCP UI resources from index-based to ID-based referencing

- Replace index-based resource markers with stable resource IDs
- Update plugin to parse \ui{resourceId} format instead of \ui0
- Refactor components to use useMessagesOperations instead of useSubmitMessage
- Add ShareMessagesProvider for UI resources in share view
- Add useConversationUIResources hook for cross-turn resource lookups
- Update parsers to generate resource IDs from content hashes
- Update all tests to use resource IDs instead of indices
- Add sandbox permissions for iframe popups
- Remove deprecated MCP tool context instructions

plgodin commented on client/src/components/Chat/Messages/Content/Markdown.tsx:
```diff
@@ -9,6 +9,11 @@ import rehypeHighlight from 'rehype-highlight';
 import type { Pluggable } from 'unified';
 import { Citation, CompositeCitation, HighlightedText } from '~/components/Web/Citation';
+import {
+  mcpUIResourcePlugin,
```
I ended up creating an ESLint plugin to enforce these conventions. [PR is here.](https://github.com/danny-avila/LibreChat/pull/10165)

samuelpath commented:
@danny-avila do you think you may have time for this 2-months old PR this week? I know you must be very busy…!

@plgodin do you think you can resolve the new merge conflicts to facilitate @danny-avila's final review here?

plgodin commented:
I just rebased and updated the branch! @danny-avila we should be good for a final review and merge!

dustinhealy reviewed (commented):
@samuelpath @plgodin 

Hi all,

I've just finished my review and everything looks good to me. Once the recent Copilot comments have been addressed I will approve the changes and hand it off to Danny for a final sign off prior to merge. 

The only friction I ran into during this review was related to AI responses incorrectly presenting the UI markers (erroneously wrapping markers in LaTex delimiters leading them to be converted into math nodes before they can be handled by the mcpUIResourcePlugin, responding with individual UIResources that are meant to be rendered within a UIResourceCarousel, etc.) causing them not to be rendered properly, but that is due to the relative strength of the models I was testing with and the context provided by the tool call responses from the MCP servers, not the implementation of the mcp-ui integration itself.

My main note which can likely be handled in a future PR is that in a shared conversation, currently there is no feedback mechanism for when a user performs a UIAction on an interactable UIResource. Of course, the core of this behavior is expected since the current strategy for handleUIAction is to send a message within the conversation / call a tool, and shared conversations in their present state are read-only; thus, such a solution is not viable. However, there still should be some indication to the user that interactions will not work as intended within that shared view beyond the console logs currently occurring when something like an "Add to Cart" button is pressed. Perhaps triggering a explanatory toast within handleUIAction for this scenario would be sufficient.

Thank you for your work on this PR - very excited to get this merged in. I believe this is a feature that many users will benefit from.

dustinhealy commented on client/src/components/MCPUIResource/plugin.ts:
```diff
@@ -0,0 +1,91 @@
+
+export const UI_RESOURCE_MARKER = '\\ui';
+// Pattern matches: \ui{id1} or \ui{id1,id2,id3} and captures everything between the braces
+export const UI_RESOURCE_PATTERN = /\\ui\{([\w]+(?:,[\w]+)*)\}/g;
```
The current pattern is correct. generateResourceId() produces 10-character hex IDs (e.g., 7e0a608c64), which contain no hyphens.

dustinhealy commented on packages/api/src/mcp/parsers.ts:
```diff
@@ -1,7 +1,12 @@
 import type * as t from './types';
 
+function generateResourceId(text: string): string {
+  return crypto.createHash('sha256').update(text).digest('hex').substring(0, 10);
```
Risk of collision negligible for the scope at which resourceIds will be generated. The increased number of tokens required to be send to models if the hashes were elongated outweighs the benefit that the collision mitigation would provide.

plgodin commented:
@danny-avila @dustinhealy I removed the extra permission from the `MCPUIResource` iframe, as Copilot's concern is valid and there are ways to securely open new pages from within a tool without opening this permission. I also fixed the merge conflicts. We should be ready to go! cc. @samuelpath

liady commented:
Really great to see this merged! Really awesome work @samuelpath ,@dustinhealy and the team 👏
Note that the official [MCP Apps spec](https://blog.modelcontextprotocol.io/posts/2025-11-21-mcp-apps/) is soon to be approved, and very soon MCP-UI client SDK [will support it](https://github.com/MCP-UI-Org/mcp-ui/pull/147) seamlessly.
